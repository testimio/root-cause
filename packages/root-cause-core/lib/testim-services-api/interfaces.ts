// generated with json2ts

interface TestimExecutionExecutionRunConfig {
  parallel: number;
  browser: string;
  runnerVersion: string;
  testimBranch: string;
  canaryMode: boolean;
  source: string;
  testPlans: any[];
  testLabels: any[];
  testNames: any[];
  testIds: any[];
  testConfigs: any[];
  testConfigIds: any[];
  browserTimeout: number;
  timeout: number;
  newBrowserWaitTimeout: number;
  tunnelPort: string;
  runnerMode: string;
  sessionType: string;
  companyId: string;
  testData?: any;
  testDataTotal?: any;
  testDataIndex?: any;
  testConfig: {}; // not too interesting
}

interface TestExecution {
  testId: string;
  status: string;
  name: string;
  resultId: string;
  isTestsContainer: boolean;
  testRunOverrideStatus: string;
  config: TestimExecutionExecutionRunConfig;
  success?: boolean;
  reason?: string;
  childTestResultIds?: string[];
  startTime: number;
  endTime: number;
  parentResultId?: string;
  show?: boolean;
}

export interface TestimBackendExecutionFormatSubsetForReporting {
  runId: string;
  execution: {
    [id: string]: {
      testId: string;
      name: string;
      success?: boolean;
      resultId: string;
      parentResultId?: string;
      isTestsContainer: boolean;
    };
  };
  startTime: number;
}
export interface TestimBackendExecutionInputFormat {
  runId: string;
  projectId: string;
  labels: string;
  startTime: number;
  endTime: number;
  execution: { [id: string]: TestExecution };
  status: string;
  config: TestimExecutionExecutionRunConfig;
  resultLabels: string[];
  remoteRunId: string;
  metadata: any;
  show: boolean;
}
