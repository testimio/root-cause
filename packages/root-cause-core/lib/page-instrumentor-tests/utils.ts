import type { RootCausePage } from '../interfaces';
import { PuppeteerPageHooker } from '../PuppeteerPageHooker';
import { TestContextMemory } from '../TestContextMemory';
import { createDateMocker } from '../utils';

export async function runPageInstrumentorTest<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage,
  testCode: (wrappedPage: TPage) => Promise<void>
): Promise<SetupPageInstrumentorForTestingReturn<TPage>['hooks']> {
  const { pageInstrumentor, wrappedPage, hooks } = await setupPageInstrumentorForTesting(
    testName,
    fullTestFileName,
    page
  );

  pageInstrumentor.start();

  await testCode(wrappedPage);

  await pageInstrumentor.end({
    success: true,
  });

  return hooks;
}

interface SetupPageInstrumentorForTestingReturn<TPage extends RootCausePage> {
  wrappedPage: TPage;
  pageInstrumentor: PuppeteerPageHooker;
  hooks: {
    beforeAllHook: jest.Mock<any, any>;
    afterAllHook: jest.Mock;
    beforeEachHook: jest.Mock;
    afterEachHook: jest.Mock;
  };
}

export async function setupPageInstrumentorForTesting<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage
): Promise<SetupPageInstrumentorForTestingReturn<TPage>> {
  const testContext = new TestContextMemory(
    '/tmp',
    testName,
    `full ${testName}`,
    fullTestFileName,
    {
      screenshots: false,
      console: false,
      networkLogs: false,
      jestAssertions: false,
      html: false,
    },
    createDateMocker()
  );

  const pageInstrumentor = new PuppeteerPageHooker(testContext, page);

  const wrappedPage = pageInstrumentor.wrapWithProxy(page);

  const beforeAllHook = jest.fn();
  const afterAllHook = jest.fn();

  const beforeEachHook = jest.fn();
  const afterEachHook = jest.fn();

  pageInstrumentor.registerBeforeAllHook(beforeAllHook);
  pageInstrumentor.registerAfterAllHook(afterAllHook);
  pageInstrumentor.registerBeforeHook(beforeEachHook);
  pageInstrumentor.registerAfterHook(afterEachHook);

  return {
    wrappedPage,
    pageInstrumentor,
    hooks: {
      beforeAllHook,
      afterAllHook,
      beforeEachHook,
      afterEachHook,
    },
  };
}

const objPropToClean = ['testContext', 'proxyContext', 'rootPage', 'instrumentedFunctionResult'];
export function cleanJestMockFunctionCalls(calls: any[]): any[] {
  return calls.map((callArg) => {
    const cloned = { ...callArg };

    for (const prop of objPropToClean) {
      if (prop in cloned) {
        cloned[prop] = 'cleaned';
      }
    }

    return cloned;
  });
}
