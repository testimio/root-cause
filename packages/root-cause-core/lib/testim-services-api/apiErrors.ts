type ApiErrorCode =
  | 'E_UNAUTHENTICATED'
  | 'E_CREDENTIALS_DONT_MATCH'
  | 'E_CREATING_EXECUTION_FAILED'
  | 'E_UPLOAD_REQUEST_FAILED'
  | 'E_FILE_NOT_FOUND';
class TestimApiErrorBase extends Error {
  constructor(public code: ApiErrorCode, message: string) {
    super(message);
  }
}
export class TestimApiError extends TestimApiErrorBase {
  constructor(private originalError: Error, message: string, code: ApiErrorCode) {
    super(code, message);
  }
}
export class TestimNotFoundError extends TestimApiErrorBase {
  constructor(message: string) {
    super('E_FILE_NOT_FOUND', message);
  }
}
