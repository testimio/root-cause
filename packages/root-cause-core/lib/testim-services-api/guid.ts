'use strict';

const a = 'qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM1234567890';

export function guid(n = 10) {
  let str = '';
  for (let i = 0; i < n; i++) {
    const index = Math.floor(Math.random() * (a.length - 1));
    str += a[index];
  }
  return str;
}
