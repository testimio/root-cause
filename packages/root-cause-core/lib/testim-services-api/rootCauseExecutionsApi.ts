import {
  runConclusionToTestimExecution,
  getRunConclusionFiles,
} from './runConclusionToTestimExecution';
import type { TestimUserMetadata, GitMetadata } from './runConclusionToTestimExecution';
import { TestimExecutionsApi } from './testimExecutionsApi';
import { TestimAssetsApi } from './testimAssetsApi';
import * as path from 'path';
import { createReadStream } from 'fs';
import { TestimBackendExecutionInputFormat } from './interfaces';
import * as fse from 'fs-extra';
import type { AbortSignal } from 'abort-controller';
import { RunConclusionFile, TestResultFile } from '@testim/root-cause-types';

type ProgressCallback = (progress: number) => void;

export class RootCauseExecutionsApi {
  constructor(
    private testimExecutionsApi: TestimExecutionsApi = new TestimExecutionsApi(),
    private testimAssetsApi: TestimAssetsApi = new TestimAssetsApi()
  ) {}
  async createExecution(
    conclusionFolderPath: string,
    testimUserMetadata: TestimUserMetadata,
    progress?: ProgressCallback,
    signal?: AbortSignal
  ) {
    // eslint-disable-next-line import/no-dynamic-require
    const conclusion = require(path.join(
      conclusionFolderPath,
      'runConclusion.json'
    )) as RunConclusionFile;
    //TODO(benji)
    // zip - maybe zip multiple photos

    const filesToUpload = await getRunConclusionFiles(conclusionFolderPath);
    const arbitraryResultsFilePath = filesToUpload.find((file) => file.endsWith('results.json'));
    const gitMetadata = arbitraryResultsFilePath
      ? extractGitMetadata(arbitraryResultsFilePath)
      : undefined;
    const execution = runConclusionToTestimExecution(conclusion, testimUserMetadata, gitMetadata);
    execution.metadata.screenplayFilesUploaded = filesToUpload.length;

    await createTempResultFilesWithResultIds(filesToUpload, execution);

    function isImage(fileName: string) {
      return fileName.endsWith('.png') || fileName.endsWith('.jpg') || fileName.endsWith('.webp');
    }
    execution.metadata.screenplayImagesUploaded = filesToUpload.filter(isImage).length;

    await this.testimExecutionsApi.createExecution(execution, signal);

    let done = 0;
    const filePromises = filesToUpload.map((file) =>
      this.testimAssetsApi
        .uploadAsset({
          // @todo sync server and client
          pathInsideBucket: `.root-cause/${conclusion.runId}/${path.relative(
            conclusionFolderPath,
            file
          )}`,
          projectId: testimUserMetadata.projectId,
          asset: createReadStream(
            file.endsWith('results.json') ? `${file.substr(0, file.length - 5)}_tmp.json` : file
          ),
          signal,
        })
        .then(() => {
          done += 1;
          if (progress) {
            progress(done / filesToUpload.length);
          }
        })
    );
    await Promise.all(filePromises);
    await removeTempResultFiles(filesToUpload);
    return { rootCauseConclusion: conclusion, testimFormat: execution };
  }
}

async function createTempResultFilesWithResultIds(
  files: string[],
  execution: TestimBackendExecutionInputFormat
) {
  const executionResults = Object.keys(execution.execution).map(
    (resultId) => execution.execution[resultId]
  );

  for (let i = 0; i < files.length; i++) {
    if (files[i].endsWith('results.json')) {
      const resultFile = await fse.readJSON(files[i]);
      resultFile.resultId = executionResults.find(
        (er) => er.endTime === resultFile.metadata.endedTimestamp
      )?.resultId;
      const tempPath = `${files[i].substr(0, files[i].length - 5)}_tmp.json`;
      await fse.writeJSON(tempPath, resultFile);
    }
  }
}

async function removeTempResultFiles(files: string[]) {
  for (let i = 0; i < files.length; i++) {
    if (files[i].endsWith('results.json')) {
      const tempPath = `${files[i].substr(0, files[i].length - 5)}_tmp.json`;
      await fse.unlink(tempPath);
    }
  }
}

function extractGitMetadata(fileName: string): GitMetadata | undefined {
  try {
    const testResult = require(fileName) as TestResultFile;
    const metadata = testResult.metadata.branchInfo;
    if (!metadata) {
      return undefined;
    }
    return {
      gitBranch: metadata.branchName,
      gitCommit: metadata.commitHash,
      gitRepoUrl: process.env.GIT_URL ?? process.env.CIRCLE_REPOSITORY_URL ?? '',
    };
  } catch (e) {
    return undefined;
  }
}
