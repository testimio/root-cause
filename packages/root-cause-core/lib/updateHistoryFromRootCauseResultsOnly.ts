import {
  readRunResultsDirToMap,
  writeRunConclusion,
  readRunsHistory,
  updateRunsHistory,
  writeRunsHistory,
} from './runConclusion/runConclusionUtils';
import { constructResultDir, constructTestInvocationResultDir } from './utils';
import { FALLBACK_RUN_ID, HISTORY_RUNS_TO_RETAIN } from './consts';
import fs from 'fs-extra';
import type { RootCauseRunResultEntry } from './runConclusion/interfaces';
import type { RunConclusionFile } from '@testim/root-cause-types';

/**
 * To be used explicitly when there's no test runner
 */
export async function updateHistoryFromRootCauseResultsOnly(
  runId = FALLBACK_RUN_ID,
  projectRoot = process.cwd(),
  dateConstructor = Date
) {
  const resultsDirPath = constructResultDir(projectRoot);
  const runResultsPath = constructTestInvocationResultDir(projectRoot, runId);

  if (!(await fs.pathExists(resultsDirPath))) {
    throw new Error(
      'None existing Root Cause path, make sure you use the same projectRoot as your attach call'
    );
  }

  if (!(await fs.pathExists(runResultsPath))) {
    throw new Error(
      'None existing Root Cause runs path, make sure you use the same runId & projectRoot as your attach call'
    );
  }

  const rootCauseResults = await readRunResultsDirToMap(runResultsPath);
  const conclusion = prepareRunConclusionFromOnlyRootCauseResults(
    runId,
    dateConstructor.now(),
    rootCauseResults
  );
  await writeRunConclusion(resultsDirPath, runId, conclusion);
  const history = await readRunsHistory(resultsDirPath);
  // We may here also actually delete old files
  updateRunsHistory(history, conclusion, HISTORY_RUNS_TO_RETAIN);
  await writeRunsHistory(resultsDirPath, history);
}

export async function readHistoryFallback(
  runId = FALLBACK_RUN_ID,
  projectRoot = process.cwd(),
  dateConstructor = Date
): Promise<RunConclusionFile> {
  const rootCausePath = constructResultDir(projectRoot);
  const rootCauseRunResultsPath = constructTestInvocationResultDir(projectRoot, runId);

  if (!(await fs.pathExists(rootCausePath))) {
    throw new Error(
      'None existing Root Cause path, make sure you use the same projectRoot as your attach call'
    );
  }

  if (!(await fs.pathExists(rootCauseRunResultsPath))) {
    throw new Error(
      'None existing Root Cause runs path, make sure you use the same runId & projectRoot as your attach call'
    );
  }

  const rootCauseResults = await readRunResultsDirToMap(rootCauseRunResultsPath);
  const conclusion = prepareRunConclusionFromOnlyRootCauseResults(
    runId,
    dateConstructor.now(),
    rootCauseResults
  );

  return conclusion;
}

function prepareRunConclusionFromOnlyRootCauseResults(
  runId: string,
  timestamp: number,
  data: Map<string, RootCauseRunResultEntry>
): RunConclusionFile {
  const conclusion: RunConclusionFile = {
    runId,
    timestamp,
    tests: [...data]
      .map(([id, entry]) => ({
        id,
        name: entry.testData.metadata.testName,
        fullName: entry.testData.metadata.testFullName,
        suiteFilePath: entry.testData.metadata.fileName || __filename,
        reason: entry.testData.metadata.testEndStatus?.success
          ? undefined
          : ((entry.testData.metadata.testEndStatus?.error as any)?.message as string),
        success: entry.testData.metadata.testEndStatus?.success ?? true,
        timestamp: entry.testData.metadata.timestamp,
        endedTimestamp: entry.testData.metadata.endedTimestamp,
      }))
      .sort((a, b) => b.timestamp - a.timestamp),
  };

  return conclusion;
}
