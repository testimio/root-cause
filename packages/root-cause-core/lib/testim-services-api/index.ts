import type nodeFetch from 'node-fetch';

import { TestimAuthApi } from './testimAuthApi';
import { TestimExecutionsApi } from './testimExecutionsApi';
import { TestimAssetsApi } from './testimAssetsApi';
import { RootCauseExecutionsApi } from './rootCauseExecutionsApi';
import type { AbortSignal } from 'abort-controller';

export { TestimAuthApi } from './testimAuthApi';
export { TestimExecutionsApi } from './testimExecutionsApi';
export { TestimAssetsApi } from './testimAssetsApi';
export { RootCauseExecutionsApi } from './rootCauseExecutionsApi';

/**
 * This class is a facade over the Testim Backend API
 */
export class TestimApi {
    auth: TestimAuthApi;
    assets: TestimAssetsApi;
    executions: TestimExecutionsApi;
    executionsApi: RootCauseExecutionsApi;
    projectId?: string;
    ciToken?: string;
    constructor(baseUrl?: string, fetch?: typeof nodeFetch) {
        this.auth = new TestimAuthApi(baseUrl, fetch);
        this.assets = new TestimAssetsApi(baseUrl, this.auth, undefined, fetch);
        this.executions = new TestimExecutionsApi(baseUrl, this.auth, fetch);
        this.executionsApi = new RootCauseExecutionsApi(this.executions, this.assets);
    }
    async authenticate(projectId: string, ciToken: string, signal?: AbortSignal) {
        await this.auth.getToken(projectId, ciToken, signal);
        this.projectId = projectId;
        this.ciToken = ciToken;
    }
}
