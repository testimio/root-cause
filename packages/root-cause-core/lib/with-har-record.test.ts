import { attach } from './index';
import type { TestResultFile } from '@testim/root-cause-types';
import puppeteer from 'puppeteer';
import fs from 'fs-extra';
import path from 'path';
import { testUniqueIdentifierFromStartParams, jsonReduceNoiseReviver } from './utils';
import {
    getCleanAllPathsPrettyFormatPlugin,
    getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
    interestingPartsOfHar,
    getPostmanEchoWorkaround1PrettyFormatPlugin,
    getPostmanEchoWorkaround2PrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
import { guid } from './testim-services-api/guid';
import { updateHistoryFromRootCauseResultsOnly } from './updateHistoryFromRootCauseResultsOnly';
// eslint-disable-next-line import/no-extraneous-dependencies
import { registerJasmineReporterToGlobal, getJasmineCurrentTest } from '@testim/root-cause-jest';

describe('with har record', () => {
    jest.setTimeout(30_000);
    expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
    expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());
    expect.addSnapshotSerializer(getPostmanEchoWorkaround1PrettyFormatPlugin());
    expect.addSnapshotSerializer(getPostmanEchoWorkaround2PrettyFormatPlugin());

    let browser: puppeteer.Browser;
    let browserContext: puppeteer.BrowserContext;
    let page: puppeteer.Page;

    beforeAll(async () => {
        registerJasmineReporterToGlobal();

        // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
        // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
        // Need to revisit
        browser = await puppeteer.launch({ args: ['--no-sandbox', '--disable-setuid-sandbox'] });
    });

    afterAll(async () => {
        await browser.close();
    });

    beforeEach(async () => {
        browserContext = await browser.createIncognitoBrowserContext();
        page = await browserContext.newPage();
    });

    afterEach(async () => {
        await page.close();
        await browserContext.close();
    });

    test('with har record', async () => {
        const currentTest = getJasmineCurrentTest();

        const localRunId = guid();

        const startTestParams = {
            runId: localRunId,
            projectRoot: process.cwd(),
            fullName: currentTest.fullName,
            description: currentTest.description,
            fullSuitePath: currentTest.testPath,
        };

        const mockedDateConstructor: typeof Date = Object.create(Date);

        let nowCallsCounter = 1;
        mockedDateConstructor.now = function mockedNow() {
            return nowCallsCounter++;
        };

        const attachController = await attach({
            page,
            startTestParams,
            activeFeatures: {
                screenshots: {
                    format: 'jpeg',
                    quality: 85,
                    fullPage: false,
                },
                console: true,
                networkLogs: true,
                jestAssertions: false,
            },
        }, mockedDateConstructor);

        const {
            page: playedPage, endTest,
        } = attachController;
        // The delay + mousedown handler inside jsbin is to make sure the requests will stay under specific test
        // Still, it might be flaky
        await playedPage.goto('http://jsbin.testim.io/ces');
        await playedPage.click('[data-job=GET_OK]', { delay: 800 });
        await playedPage.click('[data-job=GET_404]', { delay: 800 });
        await playedPage.click('[data-job=POST_OK]', { delay: 800 });
        await playedPage.click('[data-job=POST_404]', { delay: 800 });

        await endTest({
            success: true,
        });

        updateHistoryFromRootCauseResultsOnly(localRunId);

        const expectedResultsPath = path.resolve(process.cwd(), '.root-cause', 'runs', localRunId, testUniqueIdentifierFromStartParams(startTestParams));
        const testResults: TestResultFile = JSON.parse(await fs.readFile(path.resolve(expectedResultsPath, 'results.json'), 'utf-8'), jsonReduceNoiseReviver);

        expect(testResults).toMatchSnapshot('Test results.json');

        for (const stepResult of testResults.steps) {
            // Do also images blob comparison? TBH
            if (stepResult.screenshot) {
                // eslint-disable-next-line no-await-in-loop
                expect(fs.pathExists(path.resolve(expectedResultsPath, stepResult.screenshot))).resolves.toBe(true);
            }
        }

        const harFileContent = JSON.parse(await fs.readFile(path.resolve(expectedResultsPath, 'networkLogs.har'), 'utf8'));
        expect(interestingPartsOfHar(harFileContent)).toMatchSnapshot('har file content');
    });
});
