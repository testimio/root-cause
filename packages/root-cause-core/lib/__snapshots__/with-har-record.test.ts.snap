// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with har record with har record: Test results.json 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 15,
    "fileName": "noise_removed/packages/root-cause-core/lib/with-har-record.test.ts",
    "hasNetworkLogs": true,
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/85.0.4182.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "success": true,
    },
    "testFullName": "with har record with har record",
    "testName": "with har record",
    "timestamp": 2,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 4,
      "fnName": "goto",
      "index": 1,
      "name": "goto \\"http://jsbin.testim.io/ces\\"",
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 3,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.makeStep",
            "line": 235,
          },
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction",
            "line": 148,
          },
          Object {
            "column": 5,
            "file": "lib/with-har-record.test.ts",
            "function": "Object.<anonymous>",
            "line": 94,
          },
        ],
        "codeLines": Array [
          "    const { page: playedPage, endTest } = attachController;",
          "    // The delay + mousedown handler inside jsbin is to make sure the requests will stay under specific test",
          "    // Still, it might be flaky",
          "    await playedPage.goto('http://jsbin.testim.io/ces');",
          "    await playedPage.click('[data-job=GET_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=GET_404]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_OK]', { delay: 800 });",
        ],
        "column": 5,
        "fromRowNumber": 91,
        "row": 94,
        "sourceFileRelativePath": "lib/with-har-record.test.ts",
        "toRowNumber": 97,
      },
      "text": "http://jsbin.testim.io/ces",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 6,
      "fnName": "click",
      "index": 2,
      "name": "click \\"[data-job=GET_OK]\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "[data-job=GET_OK]",
      "startTimestamp": 5,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.makeStep",
            "line": 235,
          },
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction",
            "line": 148,
          },
          Object {
            "column": 5,
            "file": "lib/with-har-record.test.ts",
            "function": "Object.<anonymous>",
            "line": 95,
          },
        ],
        "codeLines": Array [
          "    // The delay + mousedown handler inside jsbin is to make sure the requests will stay under specific test",
          "    // Still, it might be flaky",
          "    await playedPage.goto('http://jsbin.testim.io/ces');",
          "    await playedPage.click('[data-job=GET_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=GET_404]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_404]', { delay: 800 });",
        ],
        "column": 5,
        "fromRowNumber": 92,
        "row": 95,
        "sourceFileRelativePath": "lib/with-har-record.test.ts",
        "toRowNumber": 98,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [
        Object {
          "args": Array [],
          "level": "error",
          "text": "Failed to load resource: the server responded with a status of 404 ()",
          "timestamp": 8,
          "url": "https://postman-echo.com/status/404",
        },
      ],
      "endTimestamp": 9,
      "fnName": "click",
      "index": 3,
      "name": "click \\"[data-job=GET_404]\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "[data-job=GET_404]",
      "startTimestamp": 7,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.makeStep",
            "line": 235,
          },
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction",
            "line": 148,
          },
          Object {
            "column": 5,
            "file": "lib/with-har-record.test.ts",
            "function": "Object.<anonymous>",
            "line": 96,
          },
        ],
        "codeLines": Array [
          "    // Still, it might be flaky",
          "    await playedPage.goto('http://jsbin.testim.io/ces');",
          "    await playedPage.click('[data-job=GET_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=GET_404]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_404]', { delay: 800 });",
          "",
        ],
        "column": 5,
        "fromRowNumber": 93,
        "row": 96,
        "sourceFileRelativePath": "lib/with-har-record.test.ts",
        "toRowNumber": 99,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 11,
      "fnName": "click",
      "index": 4,
      "name": "click \\"[data-job=POST_OK]\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "[data-job=POST_OK]",
      "startTimestamp": 10,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.makeStep",
            "line": 235,
          },
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction",
            "line": 148,
          },
          Object {
            "column": 5,
            "file": "lib/with-har-record.test.ts",
            "function": "Object.<anonymous>",
            "line": 97,
          },
        ],
        "codeLines": Array [
          "    await playedPage.goto('http://jsbin.testim.io/ces');",
          "    await playedPage.click('[data-job=GET_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=GET_404]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_404]', { delay: 800 });",
          "",
          "    await endTest({",
        ],
        "column": 5,
        "fromRowNumber": 94,
        "row": 97,
        "sourceFileRelativePath": "lib/with-har-record.test.ts",
        "toRowNumber": 100,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [
        Object {
          "args": Array [],
          "level": "error",
          "text": "Failed to load resource: the server responded with a status of 404 ()",
          "timestamp": 13,
          "url": "https://postman-echo.com/status/404",
        },
      ],
      "endTimestamp": 14,
      "fnName": "click",
      "index": 5,
      "name": "click \\"[data-job=POST_404]\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "[data-job=POST_404]",
      "startTimestamp": 12,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.makeStep",
            "line": 235,
          },
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction",
            "line": 148,
          },
          Object {
            "column": 5,
            "file": "lib/with-har-record.test.ts",
            "function": "Object.<anonymous>",
            "line": 98,
          },
        ],
        "codeLines": Array [
          "    await playedPage.click('[data-job=GET_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=GET_404]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_OK]', { delay: 800 });",
          "    await playedPage.click('[data-job=POST_404]', { delay: 800 });",
          "",
          "    await endTest({",
          "      success: true,",
        ],
        "column": 5,
        "fromRowNumber": 95,
        "row": 98,
        "sourceFileRelativePath": "lib/with-har-record.test.ts",
        "toRowNumber": 101,
      },
      "unhandledExceptions": Array [],
    },
  ],
}
`;

exports[`with har record with har record: har file content 1`] = `
Object {
  "log": Object {
    "creator": Object {
      "comment": "Created during a test executed by Testim.",
      "name": "Testim",
      "version": "1.0",
    },
    "entries": Array [
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "http://jsbin.testim.io/ces",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "http://jsbin-static.testim.io/js/render/edit.js?4.1.4",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "http://jsbin-static.testim.io/css/edit.css",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "http://jsbin-static.testim.io/images/favicon.png",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "https://postman-echo.com/status/200",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "GET",
          "postData": undefined,
          "url": "https://postman-echo.com/status/404",
        },
        "response": Object {
          "redirectURL": "",
          "status": 404,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "POST",
          "postData": undefined,
          "url": "https://postman-echo.com/post",
        },
        "response": Object {
          "redirectURL": "",
          "status": 200,
          "statusText": "",
        },
      },
      Object {
        "request": Object {
          "comment": undefined,
          "method": "POST",
          "postData": undefined,
          "url": "https://postman-echo.com/status/404",
        },
        "response": Object {
          "redirectURL": "",
          "status": 404,
          "statusText": "",
        },
      },
    ],
    "pages": Array [
      Object {
        "_URL": undefined,
        "title": "http://jsbin.testim.io/ces",
      },
    ],
    "version": "1.2",
  },
}
`;
